draft-irtf-cfrg-hpke.md

```diff
@@ -794,22 +784,51 @@ def OpenAuthPSK(enc, skR, info, aad, ct, psk, pskID, pkS):
 
 ## Key Encapsulation Mechanisms (KEMs) {#kem-ids}
 
-| Value  | KEM                            | Nzz  | Nenc | Npk | Reference                    |
-|:-------|:-------------------------------|:-----|:-----|:----|:-----------------------------|
-| 0x0000 | (reserved)                     | N/A  | N/A  | N/A | N/A                          |
-| 0x0010 | DHKEM(P-256, HKDF-SHA256)      | 32   | 65   | 65  | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0011 | DHKEM(P-384, HKDF-SHA384)      | 48   | 97   | 97  | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0012 | DHKEM(P-521, HKDF-SHA512)      | 64   | 133  | 133 | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0020 | DHKEM(X25519, HKDF-SHA256) | 32   | 32   | 32  | {{?RFC7748}}, {{?RFC5869}}   |
-| 0x0021 | DHKEM(X448, HKDF-SHA512)   | 64   | 56   | 56  | {{?RFC7748}}, {{?RFC5869}}   |
+| Value  | KEM                            | Nzz  | Nenc | Npk | Nsk | Reference                    |
+|:-------|:-------------------------------|:-----|:-----|:----|:----|:-----------------------------|
+| 0x0000 | (reserved)                     | N/A  | N/A  | N/A | N/A | N/A                          |
+| 0x0010 | DHKEM(P-256, HKDF-SHA256)      | 32   | 65   | 65  | 32  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0011 | DHKEM(P-384, HKDF-SHA384)      | 48   | 97   | 97  | 48  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0012 | DHKEM(P-521, HKDF-SHA512)      | 64   | 133  | 133 | 66  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0020 | DHKEM(X25519, HKDF-SHA256) | 32   | 32   | 32  | 32  | {{?RFC7748}}, {{?RFC5869}}   |
+| 0x0021 | DHKEM(X448, HKDF-SHA512)   | 64   | 56   | 56  | 56  | {{?RFC7748}}, {{?RFC5869}}   |
+
+### Marshal/Unmarshal
+
+For the NIST curves P-256, P-384 and P-521, the Marshal function of the
+KEM performs the uncompressed Elliptic-Curve-Point-to-Octet-String
+conversion according to {{SECG}}. The Unmarshal function performs the
+uncompressed Octet-String-to-Elliptic-Curve-Point conversion.
```

Comment:
The Unmarshal function defined by SECG can return the point at infinity. Worth considering if you want to explicitly disallow that.

Reply:
This is what the current spec does, I'm just more explicit about it here.

That said, I can share some thoughts.
* The Unmarshal procedure should not really be used for deriving keys, but for reconstructing saved keys from a previous session. If Unmarshal produces a bad key, then that means you somehow generated a bad key in a previous session or the key was tampered with.
* The only way to generate a bad keypair is by giving to DeriveKeypair 0 or p (in the case of X25519), or an HKDF preimage of 0 (in the case of P- curves)
* A point at infinity must be caught at the DH stage anyway

Would it make sense to add an HKDF step to the X25519 DeriveKeyPair? It would be nice to claim that it's computationally infeasible to derive a bad keypair using DeriveKeyPair

Reply:
I'm worried about an adversary sending the point at infinity maliciously. MLS previously disallowed infinity, but now it won't

Reply:
Are you specifically concerned about an adversary giving the point-at-infinity as a DH pubkey? Doesn't the "Validation of Inputs and Outputs" section below the "DeriveKeyPair" section cover this?

Reply:
Yes. Ah, I didn't see that before! Why is it "one of the following checks" and not "both"?

Reply:
The "full public-key validation on all public key inputs" includes a point-at-infinity check and a valid-point check. If you're guaranteed that the input point is valid and has prime order (this follows from non-identity because the P- curves are all prime order), then the two checks mentioned are equivalent.

Reply:
That's not clear to me. Probably worth re-iterating that "full validation" means "coordinates are field elements, point is on curve, and point is not infinity"